columnflow/selection/stats.py

```diff
@@ -0,0 +1,169 @@
+    sequence of tuples whose elements should be names of groups in *group_names*. As the name
+    suggests, combinations of all possible values between groups are evaluated and stored in a
+    nested dictionary.
+
+    Example:
+
+    .. code-block:: python
+
+        group_combinations = [("process", "njet")]
+
+    In this case, *stats* will obtain additional fields ``"sum_mc_weight_per_process_and_njet"``
+    and ``"sum_mc_weight_selected_per_process_and_njet"``, referring to nested dictionaries whose
+    structure depends on the exact order of group names per tuple.
+    """
+    # get the plain event mask
+    event_mask = results.main.event
+
+    # increment plain counts
+    stats["n_events"] += len(events)
+    stats["n_events_selected"] += ak.sum(event_mask, axis=0)
+
+    # make values in group map unique
+    unique_group_values = {
+        np.unique(group_data["values"])
+        for group_name, group_data in group_map.items()
+    }
+
+    # treat groups as combinations of a single group
+    for group_name, group_data in list(group_map.items())[::-1]:
+        if group_data["combinations_only"] or (group_name,) in group_combinations:
+            continue
+        group_combinations.insert(0, (group_name,))
+
+    # prepare default types (nested defaultdict's) down to a certain depth
+    group_defaults = {
+        1: (lambda: defaultdict(float)),
+    }
+    for i in range(2, max([len(group_names) for group_names in group_combinations] + [0]) + 1):
+        # use a self-executing closure to avoid reliance inside the lambda on i in the loop body
+        group_defaults[i] = (lambda i: (lambda: group_defaults[i - 1]))(i)
+
+    # get and store the weights per entry in the map
+    for name, weights in weight_map.items():
+        mask = Ellipsis
+        if isinstance(weights, (tuple, list)) and len(weights) == 2:
+            weights, mask = weights
+
+        # sum for all processes
+        stats[f"sum_{name}"] += ak.sum(weights[mask])
+
+        # when mask is an Ellipsis, it cannot be & joined to other masks, so use True instead
+        joinable_mask = True if mask is Ellipsis else mask
+
+        # per group combination
+        for group_names in group_combinations:
+            group_key = f"sum_{name}_per_" + "_and_".join(group_names)
+
+            # set the default structure
+            if group_key not in stats:
+                stats[group_key] = group_defaults[len(group_names)]()
+
+            # set values
+            for values in itertools.product(*(unique_group_values[g] for g in group_names)):
+                # evaluate and join the masks
+                group_mask = reduce(
+                    and_,
+                    (group_map[g]["mask_fn"](v) for g, v in zip(group_names, values)),
+                )
+                # find the innermost dictionary to perform the in-place item assignment
+                innermost_dict = reduce(
```

Comment:
Without the `defaultdict` in the nested lambda above, this raises a `TypeError` for me for all group combinations at depth >=2:
```
TypeError: 'function' object is not subscriptable
```